### PR TITLE
DynamoDB: Return slowdown kv error

### LINKED ISSUE #7544: S3 gateway & DynamoDB store: slowdown
* Propagate the slowdown response back to the S3 gateway client.
* Return `kv.ErrSlowDown` from the DynamoDB store when [appropriate](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.Errors.html).

### PR DESCRIPTION
**Pull request by Jonathan-Rosenberg**
Closes #7544

### CODE REVIEW

File: pkg/kv/dynamodb/store.go [hunk 1]
```diff
@@ -322,6 +322,11 @@ func (s *Store) setWithOptionalPredicate(ctx context.Context, partitionKey, key,
 		if usePredicate && errors.As(err, &errConditionalCheckFailed) {
 			return kv.ErrPredicateFailed
 		}
+		var errRequestLimitExceeded *types.RequestLimitExceeded
+		var errProvisionedThroughputExceededException *types.ProvisionedThroughputExceededException
+		if errors.As(err, &errRequestLimitExceeded) || errors.As(err, &errProvisionedThroughputExceededException) {
+			return kv.ErrSlowDown
```
Comment (itaiad200) [inline pkg/kv/dynamodb/store.go:328]: This propagates to openapi controller too. Do we change the returned status code or anything else in the response as a result?

File: pkg/kv/kvtest/store.go [hunk 2]
```diff
@@ -50,6 +50,21 @@ func sampleEntry(prefix string, n int) kv.Entry {
 	return kv.Entry{Key: []byte(k), Value: []byte(v)}
 }
 
+type ExtraTest struct {
+	Name string
+	Fn   func(t *testing.T)
+}
+
+func DriverTestWithExtras(t *testing.T, ms MakeStore, extras ...ExtraTest) {
```
[the remark this answers is not in the document]
Comment (Jonathan-Rosenberg) [reply pkg/kv/kvtest/store.go:58]: Run the default tests + extra tests that are specific for the used store implementation

File: pkg/kv/dynamodb/store.go (hunk 1 above)
Comment (Jonathan-Rosenberg) [reply pkg/kv/dynamodb/store.go:328]: No, it returns an internal server error

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Mostly worried how this will be tested.

1. I am not sure that we always (or ever...) receive the errors that this PR tries to see.  AFAIK "throttling" is not an error, it is merely a slow response.  In which case this PR will do nothing.

   So I want logs / metrics that will allow us to understand what _really_ happens.
1. Does S3AFileSystem handle slow down responses?  (I am more sure that it does this, of course!)
1. Sending a "slow down" response does not give the caller much information about how to slow down.  I worry that this will be a slow recovery.

   Suppose that I am wrong on all the technical details: we get an error (and not slow service), and it is of the right type, and S3A knows to back off when it receives a "slow down" error.  Now I am over quota X, and DynamoDB gives the expected error.  Then every request I make will return "slow down".  But each pending operation still needs to be processed!  And the only way to get a "slow down" is to perform the operation and be over the quota.  So recovery will be slow: First every pending operation has to be tried on lakeFS and fail.  Only then will the client back off.


I would like to propose splitting this into 2 parts: Detection and defence.  _Detection_ means determining that we are being slowed down by DynamoDB, or that we might be slowed down by it soon.  _Defence_ means protecting ourselves from this.  This PR mixes both tasks, I would like to detect separately from defence.  So I am *requesting* at least that we detect explicitly: give a metric or a log for when lakeFS believes that it has detected an overload.  I am blocking only on this change.

File: pkg/kv/dynamodb/store.go (hunk 1 above)
Comment (N-o-Z) [reply pkg/kv/dynamodb/store.go:328]: @Jonathan-Rosenberg at the time Barak added http error 420 to all endpoints. We need to handle this in the controller and set the right code for this error.

**Review by N-o-Z — COMMENTED** [review]
Hi,
I'm not very clear on what is the purpose of the PR.
If we want to return slowdown - why are we doing this only in the gateway only for putObject?
We also need to propagate this to the client (when not using the gateway) as 420.
I also tend to agree with @arielshaqed on the Detect + Defence (though I contest the spelling...)

**Pull request closed without merging**